src/hotspot/share/ci/ciEnv.cpp

```diff
@@ -1868,11 +1869,20 @@ InstanceKlass::ClassState ciEnv::compute_init_state_for_precompiled(InstanceKlas
             KlassTrainingData* ktd = ctd->init_dep(i);
             if (ktd->has_holder() && (ktd->holder() == ik)) {
               log_trace(precompile)("%d: init_dependency: %s: %s", task()->compile_id(), InstanceKlass::state2name(ik->init_state()), ik->external_name());
-              return InstanceKlass::ClassState::fully_initialized;; // init dependency present
+              return InstanceKlass::ClassState::fully_initialized; // init dependency present
             }
           }
         }
       }
+
+      // Core java/lang/invoke classes are peculiar. They include LF invokers, which
+      // are initialized in production run, but can be non-initialized in assembly.
+      // CI query should report their status as if in production run, otherwise AOT
```

Comment:
Why would we not resolve this disparity by ensuring they are initialized in the assembly run? rather than just reporting them as initialized for production?

Reply:
I suspect it is a part of keeping assembly phase clean and resulting dump unencumbered, but I have not checked this thoroughly. So instead of figuring which condition to check in `HeapShared::initialize_java_lang_invoke()`, I opted to go conservative.

Reply:
Oh wait, and actually, this code is answering all the queries about A4. So if we do not answer here based on dependencies, we reply "don't know about this class", and the hell breaks loose. So initializing JLI classes in assembly phase would not actually help. I confused myself, because I tried to add JLI classes as artificial dependencies, and could not, because they are filtered as not loaded. So it is easier to just reply to these queries right here.

Reply:
Ok,so maybe the comment could be tweaked to reflect that?